DataBiosphere/analysis_pipeline_WDL
# Stop drop(ing chr x) and roll

This only affects the association aggregate pipeline.

**Features**
* Resolves #66 once merged
* Debug prints are now a workflow-level variable instead of having to be set for every task 
* Added a truth file for chrX (only one configuration; see "testing" for justification)
* sbg_prepare_segments_1 is now closer to its CWL equivalent

**Small improvements**
* sbg_prepare_segments_1 now uses the logger package
* A note (logging level = info) about slow delocalization in sbg_prepare_segments_1 will be printed just before delocalization, along with some advice about how to potentially speed it up
* Clarify some of the truth files
* Debug prints are now more relevant (no more "hey, this thing that will always take less than a minute no matter what you throw at it took less than a minute")
* Add an sh file to easily prepare a workdir for fast testing of the prepare segments standalone Python script

**Background**
#66 has been known about for a while, but was deprioritized originally as this sort of association test is usually just run on autosomes. The recent look-over and fixing of miscellaneous bugs gave me the opportunity to track down #66, because, to be frank, it really bothered me! As an added bonus, the bug fix brings this task closer in line with the CWL.

**Testing**
* Tested and passes on Terra, with a caveat... more on that in a moment.
* A single chrX truth file has been added for position testing. Allele and weights have not had truth files generated because I only have a position truth file from back when SB was using the older Docker image that I myself am using. 
  * I will add/update truth files for all configurations once I get the situation with Docker images sorted out with Stephanie (TLDR it seems GitHub hasn't updated with what's on Docker Hub), which will probably be in another PR.

**What's the caveat? (technical information that is probably not relevant for reviewers, but it does deserve to be documented)**
My testing process was essentially:
1. Ensure chrX output is being generated in the local_fast configuration
2. Grab truth file from a seven bridges run from six months ago, where I did confirm the CWL doesn't drop chr X in the position configuration 
3. Add that truth file to the checker's position configuration (and associated input files needed to make sure the whole thing doesn't crash)
4. Run the checker on Terra
5. Get a weird error
6. Contact Terra support; neither of us know what's going on
7. Rerun exactly what I did in step 4 twice, once with call cacheing and once without
8. Do not get the error

Here's the weird error: A single shard of assoc_aggregate task failed in the allele_run configuration, due to what appears to be a transient fumble with localization of the null_model_file. The error was that null_size couldn't be calculated from the runtime inputs, but null_size is a property of size(null_model_file), and null_model_file is a required input, and no other shards failed, and null_model_file is not being scattered upon so there should be no way it would localize in some shards but not others. Terra Support is looking into this, but it seems to have been a transient hiccup, not something I can really control.


Edit: Also resolves #67
